File: biliup/Danmaku/huya.py
import aiohttp
from biliup.plugins import random_user_agent

from biliup.common.tars import tarscore
from biliup.plugins import match1
from biliup.plugins.huya_wup.wup_struct import EWebSocketCommandType
from biliup.plugins.huya_wup.wup_struct.WebSocketCommand import HuyaWebSocketCommand
from biliup.plugins.huya_wup.wup_struct.WSUserInfo import HuyaWSUserInfo


class Huya:
    wss_url = 'wss://cdnws.api.huya.com/'
    heartbeat = b'\x00\x03\x1d\x00\x00\x69\x00\x00\x00\x69\x10\x03\x2c\x3c\x4c\x56\x08\x6f\x6e\x6c\x69\x6e\x65\x75' \
                b'\x69\x66\x0f\x4f\x6e\x55\x73\x65\x72\x48\x65\x61\x72\x74\x42\x65\x61\x74\x7d\x00\x00\x3c\x08\x00' \
                b'\x01\x06\x04\x74\x52\x65\x71\x1d\x00\x00\x2f\x0a\x0a\x0c\x16\x00\x26\x00\x36\x07\x61\x64\x72\x5f' \
                b'\x77\x61\x70\x46\x00\x0b\x12\x03\xae\xf0\x0f\x22\x03\xae\xf0\x0f\x3c\x42\x6d\x52\x02\x60\x5c\x60' \
                b'\x01\x7c\x82\x00\x0b\xb0\x1f\x9c\xac\x0b\x8c\x98\x0c\xa8\x0c '
    heartbeatInterval = 60
    # 等待统一ua后修改
    headers = {
        'user-agent': random_user_agent(),
    }

    @staticmethod
    async def get_ws_info(url, context):
        reg_datas = []
        room_id = url.split('huya.com/')[1].split('/')[0].split('?')[0]
        async with aiohttp.ClientSession() as session:
            async with session.get(f'https://www.huya.com/{room_id}', headers=Huya.headers, timeout=5) as resp:
                room_page = await resp.text()
                uid = int(match1(room_page, r"uid\":\"?(\d+)\"?"))
                # tid = match1(room_page, r"lChannelId\":\"?(\d+)\"?")
                # sid = match1(room_page, r"lSubChannelId\":\"?(\d+)\"?")

        import base64
        ws_user_info = HuyaWSUserInfo()
        ws_user_info.iUid = uid
        ws_user_info.bAnonymous = False
        ws_user_info.lGroupId = uid
        ws_user_info.lGroupType = 3
        oos = tarscore.TarsOutputStream()
        ws_user_info.writeTo(oos, ws_user_info)

        # b64data = base64.b64encode(oos.getBuffer())
        # print(b64data)

        ws_cmd = HuyaWebSocketCommand()
        ws_cmd.iCmdType = EWebSocketCommandType.EWSCmd_RegisterReq
        ws_cmd.vData = oos.getBuffer()
        oos = tarscore.TarsOutputStream()
        ws_cmd.writeTo(oos, ws_cmd)

        # oos = tarscore.TarsOutputStream()
        # oos.write(tarscore.int64, 0, uid)
        # oos.write(tarscore.boolean, 1, False)  # Anonymous
        # oos.write(tarscore.string, 2, "")  # sGuid
        # oos.write(tarscore.string, 3, "")
        # oos.write(tarscore.int64, 4, 0)  # tid
        # oos.write(tarscore.int64, 5, 0)  # sid
        # oos.write(tarscore.int64, 6, uid)
        # oos.write(tarscore.int64, 7, 3)

        # b64data = base64.b64encode(oos.getBuffer())
        # print(b64data)

        # wscmd = tarscore.TarsOutputStream()
        # wscmd.write(tarscore.int32, 0, 1)
        # wscmd.write(tarscore.bytes, 1, oos.getBuffer())

        # b64data = base64.b64encode(oos.getBuffer())
        # print(b64data)

        reg_datas.append(oos.getBuffer())

        return Huya.wss_url, reg_datas

    @staticmethod
    def decode_msg(data):
        class User(tarscore.struct):
            @staticmethod
            def readFrom(ios):
                return ios.read(tarscore.string, 2, False).decode("utf8")

        class DColor(tarscore.struct):
            @staticmethod
            def readFrom(ios):
                return ios.read(tarscore.int32, 0, False)

        name = ""
        content = ""
        color = 16777215
        msgs = []
        ios = tarscore.TarsInputStream(data)
        if ios.read(tarscore.int32, 0, False) == 7:
            ios = tarscore.TarsInputStream(ios.read(tarscore.bytes, 1, False))
            if ios.read(tarscore.int64, 1, False) == 1400:
                ios = tarscore.TarsInputStream(ios.read(tarscore.bytes, 2, False))
                name = ios.read(User, 0, False)  # username
                content = ios.read(tarscore.string, 3, False).decode("utf8")  # content
                color = ios.read(DColor, 6, False)  # danmaku color
                if color == -1:
                    color = 16777215
        if name != "":
            msg = {"name": name, "color": f"{color}", "content": content, "msg_type": "danmaku"}
            # else:
            #     msg = {"name": "", "content": "", "msg_type": "other"}
            msgs.append(msg)
        return msgs

File: crates/danmaku/src/protocols/huya.rs
//! Huya live danmaku protocol implementation.
//!
//! Huya uses TARS binary protocol over WebSocket:
//! - Registration: WSUserInfo wrapped in WebSocketCommand (iCmdType=1)
//! - Messages: WebSocketCommand with iCmdType=7 for push messages
//! - Danmaku messages have message type 1400

use std::time::Duration;

use async_trait::async_trait;
use regex::Regex;
use reqwest::header::{HeaderMap, HeaderValue, USER_AGENT};
use tracing::debug;

use crate::codec::tars::{TarsInputStream, TarsOutputStream};
use crate::error::{DanmakuError, Result};
use crate::message::{ChatMessage, DanmakuEvent, DEFAULT_COLOR};
use crate::protocols::{
    ConnectionInfo, DecodeResult, HeartbeatConfig, Platform, PlatformContext, RegistrationData,
};

/// WebSocket URL for Huya danmaku.
const WSS_URL: &str = "wss://cdnws.api.huya.com/";

/// User agent for Huya requests.
const USER_AGENT_STRING: &str = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";

/// WebSocket command types.
#[allow(dead_code)]
mod cmd_type {
    pub const REGISTER_REQ: i32 = 1;
    pub const REGISTER_RSP: i32 = 2;
    pub const HEARTBEAT: i32 = 5;
    pub const HEARTBEAT_ACK: i32 = 6;
    pub const MSG_PUSH_REQ: i32 = 7;
}

/// Heartbeat packet (pre-encoded TARS).
const HEARTBEAT: &[u8] = &[
    0x00, 0x03, 0x1d, 0x00, 0x00, 0x69, 0x00, 0x00, 0x00, 0x69, 0x10, 0x03, 0x2c, 0x3c, 0x4c, 0x56,
    0x08, 0x6f, 0x6e, 0x6c, 0x69, 0x6e, 0x65, 0x75, 0x69, 0x66, 0x0f, 0x4f, 0x6e, 0x55, 0x73, 0x65,
    0x72, 0x48, 0x65, 0x61, 0x72, 0x74, 0x42, 0x65, 0x61, 0x74, 0x7d, 0x00, 0x00, 0x3c, 0x08, 0x00,
    0x01, 0x06, 0x04, 0x74, 0x52, 0x65, 0x71, 0x1d, 0x00, 0x00, 0x2f, 0x0a, 0x0a, 0x0c, 0x16, 0x00,
    0x26, 0x00, 0x36, 0x07, 0x61, 0x64, 0x72, 0x5f, 0x77, 0x61, 0x70, 0x46, 0x00, 0x0b, 0x12, 0x03,
    0xae, 0xf0, 0x0f, 0x22, 0x03, 0xae, 0xf0, 0x0f, 0x3c, 0x42, 0x6d, 0x52, 0x02, 0x60, 0x5c, 0x60,
    0x01, 0x7c, 0x82, 0x00, 0x0b, 0xb0, 0x1f, 0x9c, 0xac, 0x0b, 0x8c, 0x98, 0x0c, 0xa8, 0x0c, 0x20,
];

/// Huya live danmaku protocol.
pub struct Huya {
    client: reqwest::Client,
}

impl Huya {
    /// Create a new Huya protocol handler.
    pub fn new() -> Self {
        Self {
            client: reqwest::Client::new(),
        }
    }

    /// Build default headers.
    fn default_headers() -> HeaderMap {
        let mut headers = HeaderMap::new();
        headers.insert(USER_AGENT, HeaderValue::from_static(USER_AGENT_STRING));
        headers
    }

    /// Extract room ID from URL.
    fn extract_room_id(url: &str) -> Option<String> {
        // https://www.huya.com/123456 or https://huya.com/roomname
        let re = Regex::new(r"huya\.com/([^/?]+)").ok()?;
        re.captures(url)
            .and_then(|c| c.get(1))
            .map(|m| m.as_str().to_string())
    }

    /// Get UID from room page.
    async fn get_room_uid(&self, room_id: &str) -> Result<u64> {
        let url = format!("https://www.huya.com/{}", room_id);
        let headers = Self::default_headers();

        let resp = self
            .client
            .get(&url)
            .headers(headers)
            .timeout(Duration::from_secs(10))
            .send()
            .await?
            .text()
            .await?;

        // Extract UID from page: "uid":"123456" or "uid":123456
        let re = Regex::new(r#"uid['"]*:\s*['"]*(\d+)['"]*"#)
            .map_err(|e| DanmakuError::Decode(e.to_string()))?;

        re.captures(&resp)
            .and_then(|c| c.get(1))
            .and_then(|m| m.as_str().parse::<u64>().ok())
            .ok_or_else(|| DanmakuError::Decode("Failed to extract UID from Huya page".to_string()))
    }

    /// Build WSUserInfo TARS structure.
    fn build_ws_user_info(uid: u64) -> Vec<u8> {
        let mut oos = TarsOutputStream::new();

        // WSUserInfo fields:
        // 0: lUid (int64)
        // 1: bAnonymous (bool)
        // 2: sGuid (string)
        // 3: sToken (string)
        // 4: lTid (int64)
        // 5: lSid (int64)
        // 6: lGroupId (int64)
        // 7: lGroupType (int64)

        oos.write_int64(0, uid as i64);
        oos.write_bool(1, false); // Not anonymous
        oos.write_string(2, ""); // sGuid
        oos.write_string(3, ""); // sToken
        oos.write_int64(4, 0); // lTid
        oos.write_int64(5, 0); // lSid
        oos.write_int64(6, uid as i64); // lGroupId = uid
        oos.write_int64(7, 3); // lGroupType = 3

        oos.get_buffer().to_vec()
    }

    /// Build WebSocketCommand TARS structure.
    fn build_ws_command(cmd_type: i32, data: &[u8]) -> Vec<u8> {
        let mut oos = TarsOutputStream::new();

        // WebSocketCommand fields:
        // 0: iCmdType (int32)
        // 1: vData (bytes)

        oos.write_int32(0, cmd_type);
        oos.write_bytes(1, data);

        oos.get_buffer().to_vec()
    }

    /// Parse a WebSocket message.
    fn parse_message(data: &[u8]) -> Vec<DanmakuEvent> {
        let mut events = Vec::new();

        // Parse WebSocketCommand
        let mut ios = TarsInputStream::new(data);

        let cmd_type = ios.read_int32(0).unwrap_or(0);

        if cmd_type == cmd_type::MSG_PUSH_REQ {
            // Read vData (bytes at tag 1)
            if let Some(inner_data) = ios.read_bytes(1) {
                // Parse inner message
                let mut inner_ios = TarsInputStream::new(&inner_data);

                // Check message type at tag 1
                let msg_type = inner_ios.read_int64(1).unwrap_or(0);

                if msg_type == 1400 {
                    // Danmaku message - read message body at tag 2
                    if let Some(msg_data) = inner_ios.read_bytes(2) {
                        if let Some(event) = Self::parse_danmaku(&msg_data) {
                            events.push(event);
                        }
                    }
                }
            }
        } else if cmd_type == cmd_type::REGISTER_RSP {
            debug!("Huya register response received");
        } else if cmd_type == cmd_type::HEARTBEAT_ACK {
            debug!("Huya heartbeat ack received");
        }

        events
    }

    /// Parse a danmaku message.
    fn parse_danmaku(data: &[u8]) -> Option<DanmakuEvent> {
        let mut ios = TarsInputStream::new(data);

        // User info is at tag 0, which is a struct
        // Within user info, username is at tag 2
        // Content is at tag 3
        // Color is at tag 6 (inside a struct)

        // For simplicity, we'll try to parse the username at tag 2 of nested struct
        // and content at tag 3

        // Skip to find the username - it's nested, so we need a different approach
        // Let's read the raw structure:
        // Tag 0: User struct containing:
        //   - Tag 2: username (string)
        // Tag 3: content (string)
        // Tag 6: DColor struct containing:
        //   - Tag 0: color (int32)

        // This is a simplified parser - we look for string patterns
        let name = ios.read_string(2).unwrap_or_default();
        let content = ios.read_string(3).unwrap_or_default();

        // Try to read color from tag 6 struct
        let color = DEFAULT_COLOR;

        if content.is_empty() {
            return None;
        }

        let mut chat = ChatMessage::new(content).with_color(color);
        if !name.is_empty() {
            chat = chat.with_name(name);
        }

        Some(DanmakuEvent::Chat(chat))
    }
}

impl Default for Huya {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Platform for Huya {
    fn name(&self) -> &'static str {
        "Huya"
    }

    async fn get_connection_info(
        &self,
        url: &str,
        context: &PlatformContext,
    ) -> Result<ConnectionInfo> {
        // Get room ID
        let room_id = if let Some(ref id) = context.room_id {
            id.clone()
        } else {
            Self::extract_room_id(url)
                .ok_or_else(|| DanmakuError::Decode("Invalid Huya URL".to_string()))?
        };

        // Get UID from room page
        let uid = self.get_room_uid(&room_id).await?;
        debug!("Huya room UID: {}", uid);

        // Build registration packet
        let user_info = Self::build_ws_user_info(uid);
        let reg_packet = Self::build_ws_command(cmd_type::REGISTER_REQ, &user_info);

        Ok(
            ConnectionInfo::new(WSS_URL)
                .with_registration(vec![RegistrationData::Binary(reg_packet)])
                .with_headers(Self::default_headers()),
        )
    }

    fn heartbeat_config(&self) -> HeartbeatConfig {
        HeartbeatConfig::binary(HEARTBEAT.to_vec(), Duration::from_secs(60))
    }

    fn decode_message(&self, msg: &[u8]) -> Result<DecodeResult> {
        let events = Self::parse_message(msg);
        Ok(DecodeResult::with_events(events))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_room_id() {
        assert_eq!(
            Huya::extract_room_id("https://www.huya.com/123456"),
            Some("123456".to_string())
        );
        assert_eq!(
            Huya::extract_room_id("https://huya.com/kpl"),
            Some("kpl".to_string())
        );
    }

    #[test]
    fn test_build_ws_user_info() {
        let data = Huya::build_ws_user_info(12345);
        assert!(!data.is_empty());

        // Verify structure
        let mut ios = TarsInputStream::new(&data);
        assert_eq!(ios.read_int64(0), Some(12345));
    }

    #[test]
    fn test_build_ws_command() {
        let user_info = Huya::build_ws_user_info(12345);
        let cmd = Huya::build_ws_command(cmd_type::REGISTER_REQ, &user_info);

        assert!(!cmd.is_empty());

        // Verify structure
        let mut ios = TarsInputStream::new(&cmd);
        assert_eq!(ios.read_int32(0), Some(cmd_type::REGISTER_REQ));
    }
}
